src/core/tokenizer.c

```diff
@@ -149,31 +154,83 @@ tokenset_resolve(tokenset_t *tset, nxs_index_t *idx, bool stage)
 	}
 }
 
+
+void		char_from_uchar_slice(char *, UChar* , int32_t , int32_t);
+void
+char_from_uchar_slice(char* buf, UChar* str, int32_t start, int32_t end)
+{
+	UChar   savedEndChar;
+
+	savedEndChar = str[end];
+	str[end] = 0;
+	u_austrcpy(buf, str+start);
+	str[end] = savedEndChar;
+}
+
+
 /*
- * tokenize: split the text into tokens separated by a whitespace
- * and pass the token through the filter pipeline.
+ * tokenize: Uses ICU segmentation UBRK_WORD
+ * See: https://unicode.org/reports/tr29/
  */
 tokenset_t *
-tokenize(filter_pipeline_t *fp, const char *text, size_t text_len __unused)
+tokenize(filter_pipeline_t *fp, const char *text, size_t text_len, const char *lang)
 {
-	const char *sep = ",.;:| \t\n";
-	char *content, *val, *brk;
-	tokenset_t *tset;
+	UBreakIterator *it_token;
+	UChar       *utext = NULL;
+	UErrorCode     ec = U_ZERO_ERROR;
+	int32_t	end;
+	int32_t	start;
+	tokenset_t	*tset;
+	int32_t	ulen;
 
 	if ((tset = tokenset_create()) == NULL) {
 		return NULL;
 	}
-	if ((content = strdup(text)) == NULL) {
+
+	ulen = roundup2((text_len + 1) * 3, 64);
+	if ((utext = malloc(ulen * 2)) == NULL) {
 		goto err;
 	}
+	// if (utf8_to_utf16(NULL, text, utext, text_len) == -1) {
+	// 	goto err;
+	// }
+	u_uastrcpy(utext, text);
 
-	for (val = strtok_r(content, sep, &brk); val;
-	    val = strtok_r(NULL, sep, &brk)) {
+	/*
+	*TODO: Use word brake rules to customize.
+	* see: https://unicode-org.github.io/icu/userguide/boundaryanalysis/break-rules.html
+	*/ 
+	it_token = ubrk_open(UBRK_WORD, lang, utext, -1, &ec);
+	if (__predict_false(U_FAILURE(ec))) {
+		const char *errmsg __unused = u_errorName(ec);
+		app_dbgx("ubrk_open() failed: %s", errmsg);
+		goto err;
+	}
+
+	start = ubrk_first(it_token);
+	for (
+		end = ubrk_next(it_token); 
+		end != UBRK_DONE;
+		start = end, end = ubrk_next(it_token)
+	) {
+
+		char *val;
+		if ((val = malloc((end - start + 1) * 3 * 2)) == NULL) {
+			goto err;
+		}
+
+		char_from_uchar_slice(val, utext, start, end);
 		const size_t len = strlen(val);  // XXX
```

Comment:
This `strlen()` is redundant, your slicing function already knows the length and should return it. But the question is, why `u_strToUTF8()` / `utf8_from_utf16()` doesn't work for you?

Reply:
start and end make reference to the utext, which is u16, why end- start would work in this case?

Reply:
ignore me